service/kvmconfig/v6/key/key.go

```diff
@@ -124,18 +128,54 @@ func MasterHostPathVolumeDir(clusterID string, vmNumber string) string {
 }
 
 // MemoryQuantity returns a resource.Quantity that represents the memory to be used by the nodes.
-// It adds the memory from the node definition parameter to the additional memory defined by
-// defaultWorkerMemory.
-func MemoryQuantity(n v1alpha1.KVMConfigSpecKVMNode) (resource.Quantity, error) {
-	q, err := resource.ParseQuantity(n.Memory)
-	if err != nil {
-		return resource.Quantity{}, microerror.Maskf(err, "creating Memory quantity from node definition")
-	}
-	additionalMemory, err := resource.ParseQuantity(defaultWorkerMemory)
-	if err != nil {
-		return resource.Quantity{}, microerror.Maskf(err, "creating Memory quantity from addtional memory")
+// It adds the memory from the node definition parameter to the additional memory calculated on the node role
+func MemoryQuantity(n v1alpha1.KVMConfigSpecKVMNode, role string) (resource.Quantity, error) {
+	var q resource.Quantity
+	var err error
+	if role == MasterID {
+		q, err = resource.ParseQuantity(n.Memory)
+		if err != nil {
+			return resource.Quantity{}, microerror.Maskf(err, "creating Memory quantity from node definition")
+		}
+		additionalMemory, err := resource.ParseQuantity(baseMasterMemoryOverhead)
+		if err != nil {
+			return resource.Quantity{}, microerror.Maskf(err, "creating Memory quantity from addtional memory")
+		}
+		q.Add(additionalMemory)
+
+	} else if role == WorkerID {
+		i, err := strconv.Atoi(n.Memory[:len(n.Memory)-1])
+		if err != nil {
+			return resource.Quantity{}, microerror.Maskf(err, "calculating memory overhead multiplier")
+		}
+		// base worker memory calculated in MB
+		baseMemoryMB := strconv.Itoa(i * 1024)
+
+		q, err = resource.ParseQuantity(baseMemoryMB + "M")
+		if err != nil {
+			return resource.Quantity{}, microerror.Maskf(err, "creating Memory quantity from node definition")
+		}
+		// IO overhead for qemu is around 512M memory
+		IOOverhead, err := resource.ParseQuantity(workerIOOverhead)
+		if err != nil {
+			return resource.Quantity{}, microerror.Maskf(err, "creating Memory quantity from io overhead")
+		}
+		q.Add(IOOverhead)
+		// memory overhead is more complex as it increases with the size of the memory
+		// basic calculation is (2 + (memory / 12))*512M
+		// examples:
+		// Memory under 15G >> overhead 1024M
+		// memory between 15 - 30G >> overhead 1536M
+		// memory between 30 - 45G >> overhead 2048M
+		overheadMultiplier := int(baseWorkerOverheadMultiplier + i/baseWorkerOverheadModulator)
+		workerMemoryOverhead := strconv.Itoa(baseWorkerMemoryOverhead*overheadMultiplier) + "M"
+
+		memOverhead, err := resource.ParseQuantity(workerMemoryOverhead)
+		if err != nil {
+			return resource.Quantity{}, microerror.Maskf(err, "creating Memory quantity from memory overhead")
+		}
+		q.Add(memOverhead)
 	}
```

Comment:
It would be more type safe to split it to two separate functions like `MemoryQuantityMaster` and `MemoryQuantityWorker`. When we keep that `if` the code can fail silently because of a typo.

Reply:
splitted